dali/kernels/erase/erase_gpu.h

```diff
@@ -0,0 +1,526 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_ERASE_ERASE_GPU_H_
+#define DALI_KERNELS_ERASE_ERASE_GPU_H_
+
+#include <vector>
+
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/kernel.h"
+
+#include "dali/core/geom/vec.h"
+#include "dali/core/geom/box.h"
+#include "dali/kernels/common/utils.h"
+
+#include "dali/core/cuda_event.h"
+#include "dali/core/format.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * We divide the work in following way:
+ *
+ * We cover every n-dimensional sample by n-dimensional cubes (in fact a hyperrectangle)
+ * of one selected size. We call such a hyperrectangle a region.
+ *
+ * The size of such region is suitable for traversal of block of threads of dimension (32, 32, 1)
+ *
+ *
+ * Grid: (number of cubes, number of samples, 1)
+ * - the number of cubes is the max over the numbers of cubes to cover each sample
+ *
+ * Block: (32, 32, 1) - the dimensions of thread block are fixed. We use such "physical block"
+ * to traverse every region - the hyperrectangle.
+ *
+ * Steps:
+ * 1. In every block we first look for intersections with erased regions, to filter out only
+ * the ones that we actually need.
+ * 2. We check if the hyperrectangle is fully covered by erase region or not covered at all
+ * to delegate to special cases - fill only or memcopy
+ * 3. If we have mixed cover, we check each coordinate and either copy the input or the fill value
+ *
+ * The processing of hyperrectangle of n-dims is done by n-nested for loop from outer to innermost
+ * dimension.
+ * The two innermost dimensions (omitting the channel) are traversed by the block of parallel
+ * threads.
+ * The loop is implemented as template based recursion.
+ * Possible optimization/obfuscation - factor out checks over dimension to specific level of
+ * the loop nest (for the mixed cover case).
+ *
+ */
+
+template <typename T, int ndim>
+using box = Box<ndim, T>;
+
+template <int ndim>
+using ibox = box<int32_t, ndim>;
+
+template <typename T, int ndim>
+struct erase_sample_desc {
+  const T *in = nullptr;
+  T* out = nullptr;
+  const ibox<ndim> *erase_regions = nullptr;
+  ivec<ndim> sample_shape;
+  ivec<ndim> sample_stride;
+};
+
+template <int ndim>
+DALI_HOST_DEV ivec<ndim> get_region_cover(ivec<ndim> region_dims, ivec<ndim> sample_dims) {
```

Comment:
This is confusing... we use `dims` more as axes than actual values.
```suggestion
DALI_HOST_DEV ivec<ndim> get_region_cover(ivec<ndim> region_sizes, ivec<ndim> sample_sizes) {
```
?

Reply:
I'd rather go with shape, size feels like 1D thing.